cmd/data-usage.go

```diff
@@ -0,0 +1,205 @@
+ * limitations under the License.
+ */
+
+package cmd
+
+import (
+	"context"
+	"math/rand"
+	"time"
+)
+
+const (
+	dataUsageSweepInterval   = 12 * time.Hour
+	dataUsageRefreshInterval = 10 * time.Minute
+)
+
+// A safe copy of a data usage info struct
+func copyDataUsageInfo(dataUsage DataUsageInfo) DataUsageInfo {
+	dataUsageCopy := DataUsageInfo{}
+	dataUsageCopy.ObjectsCount = dataUsage.ObjectsCount
+	dataUsageCopy.ObjectsTotalSize = dataUsage.ObjectsTotalSize
+	dataUsageCopy.BucketsCount = dataUsage.BucketsCount
+	dataUsageCopy.LastUpdate = dataUsage.LastUpdate
+	dataUsageCopy.ObjectsSizesHistogram = make(map[string]uint64)
+	for k, v := range dataUsage.ObjectsSizesHistogram {
+		dataUsageCopy.ObjectsSizesHistogram[k] = v
+	}
+	dataUsageCopy.BucketsSizes = make(map[string]uint64)
+	for k, v := range dataUsage.BucketsSizes {
+		dataUsageCopy.BucketsSizes[k] = v
+	}
+
+	return dataUsageCopy
+}
+
+// Safe copy of the global data usage info
+func safeCopyDataUsageInfo() DataUsageInfo {
+	globalDataUsageInfoMu.RLock()
+	defer globalDataUsageInfoMu.RUnlock()
+
+	return copyDataUsageInfo(globalDataUsageInfo)
+}
+
+func initDataUsageStats() {
+	go runDataUsageInfoUpdateRoutine()
+}
+
+func runDataUsageInfoUpdateRoutine() {
+	// Wait until the object layer is ready
+	var objAPI ObjectLayer
+	for {
+		objAPI = newObjectLayerWithoutSafeModeFn()
+		if objAPI == nil {
+			time.Sleep(time.Second)
+			continue
+		}
+		break
+	}
+
+	ctx := context.Background()
+
+	zones, ok := objAPI.(*xlZones)
+	if !ok {
+		runDataUsageInfoForFS(ctx, objAPI)
+	} else {
+		runDataUsageInfoForXL(ctx, zones)
+	}
+}
+
+func runDataUsageInfoForFS(ctx context.Context, objAPI ObjectLayer) {
+	for {
+		usageInfo := objAPI.DataUsageInfo(ctx)
+
+		globalDataUsageInfoMu.Lock()
+		globalDataUsageInfo = usageInfo
+		globalDataUsageInfoMu.Unlock()
+
+		time.Sleep(dataUsageSweepInterval)
+	}
+}
+
+func runDataUsageInfoForXL(ctx context.Context, zones *xlZones) {
+	for {
+		locker := zones.NewNSLock(ctx, minioMetaBucket, "leader-data-usage-info")
+		err := locker.GetLock(newDynamicTimeout(time.Millisecond, time.Millisecond))
+		if err != nil {
+			time.Sleep(5 * time.Minute)
+			continue
+		}
+		// Break without locking
+		break
+	}
+
+	dataUsageSweepTicker := time.NewTicker(dataUsageSweepInterval)
+	dataUsageRefreshTicker := time.NewTicker(dataUsageRefreshInterval)
+
+	// Immediately ask to refresh data usage
+	zonesCallAsyncRefreshDataUsage(ctx, zones)
+
+	for {
+		select {
+		case <-dataUsageSweepTicker.C:
+			zonesCallAsyncRefreshDataUsage(ctx, zones)
+		case <-dataUsageRefreshTicker.C:
+			zonesUpdateDataUsage(ctx, zones)
+		}
+	}
+}
+
+// Choose one disk in each set in each zone and asks it to refresh its data usage.
+func zonesCallAsyncRefreshDataUsage(ctx context.Context, zones *xlZones) {
+	for _, z := range zones.zones {
+		for _, xl := range z.sets {
+			var onlineDisks []StorageAPI
+			for _, d := range xl.getDisks() {
```

Comment:
use getLoadBalancedDisks() to get the randomized shuffle @vadmeste

Reply:
> use getLoadBalancedDisks() to get the randomized shuffle @vadmeste

Good idea, done!